plasma/models/mpi_runner.py

```diff
@@ -792,19 +827,74 @@ def mpi_make_predictions(conf, shot_list, loader, custom_path=None):
+                max_length = max([max(y.shape) for y in y_p])
+                g.write_all(' max length = {}\n'.format(max_length))
+            g.comm.Barrier()
+            max_length = g.comm.allreduce(max_length, MPI.MAX) 
+            if color == 1:
+                y_prime_numpy = np.stack([np.pad(sublist, pad_width=((0,max_length-max(sublist.shape)),(0,0))) for sublist in y_prime])
+                y_gold_numpy = np.stack([np.pad(sublist, pad_width=((0,max_length-max(sublist.shape)),(0,0))) for sublist in y_gold])
+            
+            g.comm.Barrier()
+            temp_predictor_only_comm = MPI.Comm.Split(g.comm, color, i)
             g.comm.Barrier()
-            y_prime_global += concatenate_sublists(g.comm.allgather(y_prime))
-            y_gold_global += concatenate_sublists(g.comm.allgather(y_gold))
+            # Create numpy array to store all processors output, then aggregate and unpad using MPI gathered shape list
+            shpzg = g.comm.allgather(shpz)
+            shpzg = list(itertools.chain(*shpzg))
+            g.comm.Barrier()
+            shpzg = [s for s in shpzg if s != []]
+            max_length = g.comm.allreduce(max_length, MPI.MAX) 
+            if color == 1:
+                num_pred = temp_predictor_only_comm.size
+            else:
+                num_pred = g.comm.size - temp_predictor_only_comm.size
+            y_primeg = np.zeros((num_pred*conf['model']['pred_batch_size'],max_length,1), dtype=conf['data']['floatx'])
+            y_goldg  = np.zeros((num_pred*conf['model']['pred_batch_size'],max_length,1), dtype=conf['data']['floatx'])
+            y_primeg_flattend = np.zeros(y_primeg.flatten().shape)
+            y_goldg_flattend  = np.zeros(y_goldg.flatten().shape)
+            if color == 1:
+                # Ensure that numpy arrays have correct dimensions before gathering them
+                assert num_pred*max(y_prime_numpy.flatten().shape) == max(y_primeg_flattend.shape)
+                assert num_pred*max(y_gold_numpy.flatten().shape) == max(y_goldg_flattend.shape)
+                temp_predictor_only_comm.Barrier()
+                temp_predictor_only_comm.Allgather(y_prime_numpy.flatten(), y_primeg_flattend)
+                temp_predictor_only_comm.Allgather(y_gold_numpy.flatten(), y_goldg_flattend)
+                temp_predictor_only_comm.Barrier()
+            # Process 0 broadcast y_primeg adn y_goldg to all processors, including ones
+            # not involved in calculating predictions so they can each create their own 
+            # y_prime_global and y_gold_global
+            g.comm.Barrier()
+            g.comm.Bcast(y_primeg_flattend, root=0)
+            g.comm.Bcast(y_goldg_flattend, root=0) 
+            g.comm.Barrier()
+            y_primeg_flattend = np.split(y_primeg_flattend, num_pred)
+            y_goldg_flattend = np.split(y_goldg_flattend, num_pred)
+            y_primeg = [y.reshape((128, max_length, 1)) for y in y_primeg_flattend]
```

Comment:
What is the hardcoded 128 here? Should it be a tunable parameter from `conf.yaml`?

Reply:
This should be the batch size from the `conf.yaml`. I'll change it to that and run a test run on Traverse

Reply:
Changed and seems to work